### ISSUE TITLE
QAnon tweets: logistic regression model with TFIDF

### ISSUE THREAD

**Issue by jwzimmer-zz** [jwzimmer/aboutvsof#6]
Re discussion with Josh Minot, want to use a logistic regression model to identify some words and potentially ngrams which can be used to differentiate Q and non-Q content on twitter, purpose being to support/ enhance the words manually found.

**My noob notes**
- document embedding/ word embedding
- semantic space
- bert
- word2vec
- glove embeddings
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html#examples-using-sklearn-feature-extraction-text-tfidfvectorizer
- https://gitlab.com/compstorylab/tweet_utils/-/blob/master/twitter_utils/tweets.py
  - read_text()  
- Downloaded some initial tweets to get started from Josh
  - Do not commit tweets to repo
  - Do not commit twitter usernames to repo
- Tweets take up a lot of space; need to interact with the compressed file
- Tweets are surprisingly long json objects with a lot of information I can probably ignore
- Can save the exact version of the model as a pickled object

**Comment by jwzimmer-zz** [comment]
Started local notebook for this:  QTweetsLogisticRegressionTFIDF

**Comment by jwzimmer-zz** [comment]
Using just a small number of tweets and the body of the tweet rather than the better get_text result.

With Count (not TFIDF):
<img width="824" alt="Screen Shot 2021-01-31 at 5 29 06 PM" src="https://user-images.githubusercontent.com/1320456/106481303-4cab1400-647a-11eb-81d5-afad34135140.png">

With TFIDF (not Count):
<img width="824" alt="Screen Shot 2021-01-31 at 5 31 54 PM" src="https://user-images.githubusercontent.com/1320456/106481304-4cab1400-647a-11eb-85db-61e208deade3.png">

Here's how I'm finding and labelling "Qanon" and "not-Qanon" tweets:
```
train2 = train.copy() #so if I mess it up I don't have to re-run getting train

def label_qanon(row):
    if type(row['body']) == str:
        if bool(re.search(r'qanon|Qanon|QANON|QAnon',row['body'])):
            return 1
        else:
            return 0
    else:
        return 0

train2['qanon'] = train2.apply(lambda row: label_qanon(row), axis=1)

train2.head()
```

**Comment by jwzimmer-zz** [comment]
Meeting 2/2/21:
- "About" qanon: https://twitter.com/i/lists/1350837342622375936
  - try as "non-q" class in model
  - set up different comparison classes
- topic model with sklearn
- email chris for vacc access
- get set up on vacc - if there is a meeting about how to do this, invite Jane
- set up 2 grams
- set up 3 grams
- check out plotly: https://plotly.com/python/plotly-express/
- keep a lot of charts

**Comment by jwzimmer-zz** [comment]
Using all the tweets from the accounts that have a matching qanon tweet and TFIDF:
<img width="824" alt="Screen Shot 2021-02-01 at 7 05 37 PM" src="https://user-images.githubusercontent.com/1320456/106661582-5c585480-656f-11eb-95e0-5db8851e38d6.png">

Here's the relevant code... Here's how I'm finding and labelling the Qanon accounts and their tweets (vs. everything else is a non-Qanon tweet):
```
train3 = train.copy()

def find_qanon_accounts(df):
    qanonaccounts = []
    df = df.copy()
    for row in df.iterrows():
        if type(row[1]['body']) == str:
            if bool(re.search(r'qanon|Qanon|QANON|QAnon',row[1]['body'])):
                username = row[1]['actor']['preferredUsername']
                qanonaccounts += [username]
            else:
                pass
        else:
            pass
    return list(set(qanonaccounts))

qanonaccounts = find_qanon_accounts(train3)

def label_qanon(row):
    if row['actor']['preferredUsername'] in set(qanonaccounts):
        return 1
    else:
        return 0

train3['qanon'] = train3.apply(lambda row: label_qanon(row), axis=1)

train3.head()
```

**Comment by jwzimmer-zz** [comment]
Using 2grams as features + TFIDF + QAnon accounts:
- get runtime warning
- get error with validation set if I use transform instead of fit_transform

Training Accuracy:  0.999388402757171

<img width="401" alt="Screen Shot 2021-02-08 at 1 12 58 PM" src="https://user-images.githubusercontent.com/1320456/107263172-7088cf80-6a0f-11eb-9013-967c2e1bb28e.png">
<img width="866" alt="Screen Shot 2021-02-08 at 1 13 14 PM" src="https://user-images.githubusercontent.com/1320456/107263173-7088cf80-6a0f-11eb-89c3-5a02efe91f21.png">

**Comment by jwzimmer-zz** [comment]
<img width="993" alt="Screen Shot 2021-02-08 at 1 25 58 PM" src="https://user-images.githubusercontent.com/1320456/107264656-3ae4e600-6a11-11eb-83bc-a681ae52d95f.png">

**Comment by jwzimmer-zz** [comment]
Add list of control accounts from https://twitter.com/i/lists/1350837342622375936/members:
`write_json(controlaccts,"list_control_Q_accts_02-08-2021.json")`
<img width="888" alt="Screen Shot 2021-02-08 at 3 04 24 PM" src="https://user-images.githubusercontent.com/1320456/107274887-05df9000-6a1f-11eb-9276-65dd5c6037d3.png">

I think this is ready to run/ works, EXCEPT that none of the control accounts happen to be in the tweet sample Josh gave me, so I get an error that there is only 1 class of data (the Q account tweets).
